File: modules/data/crossfade.lua
sptbl["crossfade"] = {

    files = {
        module = "crossfade.c",
        header = "crossfade.h",
        example = "ex_crossfade.c",
    },

    func = {
        create = "ut_crossfade_create",
        destroy = "ut_crossfade_destroy",
        init = "ut_crossfade_init",
        compute = "ut_crossfade_compute",
    },

    params = {
        mandatory = {
        },

        optional = {
            {
                name = "pos",
                type = "UTFLOAT",
                description = "Crossfade position. 0 = all signal 1, 1 = all signal 2",
                default = 0.5
            },
        }
    },

    modtype = "module",

    description = [[Crossfade two signals. 
This module will perform a linear crossfade between two input signals.
]],

    ninputs = 2,
    noutputs = 1,

    inputs = {
        {
            name = "in1",
            description = "Input signal 1."
        },
        {
            name = "in2",
            description = "Input signal 2."
        },
    },

    outputs = {
        {
            name = "out",
            description = "signal output."
        },
    }

}

File: modules/crossfade.c
#include <stdlib.h>
#include "utone.h"

int ut_crossfade_create(ut_crossfade **p)
{
    *p = malloc(sizeof(ut_crossfade));
    return UT_OK;
}

int ut_crossfade_destroy(ut_crossfade **p)
{
    free(*p);
    return UT_OK;
}

int ut_crossfade_init(ut_data *ut, ut_crossfade *p)
{
    p->pos = 0.5;
    return UT_OK;
}

int ut_crossfade_compute(ut_data *ut, ut_crossfade *p, UTFLOAT *in1, UTFLOAT *in2, UTFLOAT *out)
{
    *out = *in2 * p->pos + *in1 * (1 - p->pos);
    return UT_OK;
}
